### PR TITLE
failed on unregistered featuregate

### LINKED ISSUE #5656: Collector should fail when passed an unregistered feature gate
**Describe the bug**

The Collector should fail fast when passing a feature gate that is unregistered

**Steps to reproduce**

Run the Collector while passing a nonexistent feature gate through CLI like `--feature-gates=garbage.value`.

**What did you expect to see?**

The Collector fails to start and tells me the feature gate `garbage.value` does not exist.

**What did you see instead?**

The Collector starts while ignoring the garbage value.

**What version did you use?**

v0.55.0

**Additional context**

At the Go API level, we should make [`featuregate.Apply`](https://github.com/open-telemetry/opentelemetry-collector/blob/58e66af2558a56be1f26c759db80e0ed35234607/service/featuregate/gates.go#L49) return an error when passing an unregistered feature flag.  This falls onto the [second case](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#example-2---changing-the-return-values-of-a-function) from the contributing guidelines

### PR DESCRIPTION
**Pull request by fatsheep9146**
Signed-off-by: Ziqi Zhao <zhaoziqi9146@gmail.com>

**Description:** <Describe what has changed. 
Enhance registry to support failing on unregistered featuregate

fix #5656

### CODE REVIEW

File: service/featuregate/gates.go [hunk 1]
```diff
@@ -57,6 +58,21 @@ func (r *Registry) Apply(cfg map[string]bool) {
...
+// Sets only those values provided in the map, other gate values are not changed.
+func (r *Registry) MustApply(cfg map[string]bool) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	for id, val := range cfg {
+		if g, ok := r.gates[id]; ok {
+			g.Enabled = val
+			r.gates[g.ID] = g
+		} else {
+			panic(fmt.Sprintf("feature gate %s is unregistered", id))
```
Comment (codeboten) [inline service/featuregate/gates.go:71]: Thoughts on returning an error here instead of calling `panic`? It's what @mx-psi's original suggestion was.

**Review by codeboten — COMMENTED** [review]
One question, please add a unit test as well to ensure the functionality works as expected

**Comment by TylerHelmuth** [comment]
@fatsheep9146 can you also add a new test(s) that verify the collector fails to start on an unknown feature flag

File: service/featuregate/gates.go (hunk 1 above)
Comment (fatsheep9146) [reply service/featuregate/gates.go:71]: @mx-psi You are right, I follow the example 2.

Comment (codeboten) [reply service/featuregate/gates.go:71]: Got it. Thanks for clarifying!

**Comment by fatsheep9146** [comment]
@TylerHelmuth @codeboten 
I've already added a unit test for registered and unregistered gate case.

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,7 @@
 ### 💡 Enhancements 💡
 
 - Add `linux-ppc64le` architecture to cross build tests in CI
+- Collector failed when passed an unregistered feature gate (#5660)
```
[the remark this answers is not in the document]
Comment (fatsheep9146) [reply CHANGELOG.md:8]: The entry is supplied.

**Review by TylerHelmuth — APPROVED** [review]

File: service/command.go [hunk 3]
```diff
@@ -30,7 +30,7 @@ func NewCommand(set CollectorSettings) *cobra.Command {
 		Version:      set.BuildInfo.Version,
 		SilenceUsage: true,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			featuregate.GetRegistry().Apply(gatesList)
+			featuregate.GetRegistry().MustApply(gatesList)
```
[the remark this answers is not in the document]
Comment (fatsheep9146) [reply service/command.go:33]: That makes sense, it is really a special case of changing the function :-)

[the remark this answers is not in the document]
Comment (fatsheep9146) [reply service/command.go]: Done

**Review by mx-psi — APPROVED** [review]

**Pull request merged**